File: 0682/sesión10 - soporte MSSQL/SoftInvPersistance/src/main/java/pe/edu/pucp/softinv/daoImp/AlmacenDAOImpl.java
package pe.edu.pucp.softinv.daoImp;

import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;
import pe.edu.pucp.softinv.dao.AlmacenDAO;
import pe.edu.pucp.softinv.daoImp.util.Columna;
import pe.edu.pucp.softinv.model.AlmacenesDTO;

public class AlmacenDAOImpl extends DAOImplBase implements AlmacenDAO {

    private AlmacenesDTO almacen;

    public AlmacenDAOImpl() {
        super("INV_ALMACENES");
        this.retornarLlavePrimaria = true;
        this.almacen = null;
    }

    @Override
    protected void configurarListaDeColumnas() {
        this.listaColumnas.add(new Columna("ALMACEN_ID", true, true));
        this.listaColumnas.add(new Columna("NOMBRE", false, false));
        this.listaColumnas.add(new Columna("ALMACEN_CENTRAL", false, false));
    }

    @Override
    protected void incluirValorDeParametrosParaInsercion() throws SQLException {
        this.statement.setString(1, this.almacen.getNombre());
        this.statement.setInt(2, this.almacen.getAlmacen_central() ? 1 : 0);
    }

    @Override
    protected void incluirValorDeParametrosParaModificacion() throws SQLException {
        this.statement.setString(1, this.almacen.getNombre());
        this.statement.setInt(2, this.almacen.getAlmacen_central() ? 1 : 0);
        this.statement.setInt(3, this.almacen.getAlmacenId());
    }

    @Override
    protected void incluirValorDeParametrosParaEliminacion() throws SQLException {
        this.statement.setInt(1, this.almacen.getAlmacenId());
    }

    @Override
    protected void incluirValorDeParametrosParaObtenerPorId() throws SQLException {
        this.statement.setInt(1, this.almacen.getAlmacenId());
    }

    @Override
    protected void instanciarObjetoDelResultSet() throws SQLException {
        this.almacen = new AlmacenesDTO();
        this.almacen.setAlmacenId(this.resultSet.getInt("ALMACEN_ID"));
        this.almacen.setNombre(this.resultSet.getString("NOMBRE"));
        this.almacen.setAlmacen_central(this.resultSet.getInt("ALMACEN_CENTRAL") == 1);
    }

    @Override
    protected void limpiarObjetoDelResultSet() {
        this.almacen = null;
    }

    @Override
    protected void agregarObjetoALaLista(List lista) throws SQLException {
        this.instanciarObjetoDelResultSet();
        lista.add(this.almacen);
    }

    @Override
    public Integer insertar(AlmacenesDTO almacen) {
        this.almacen = almacen;
        return super.insertar();
//        int resultado = 0;
//        try {
//            this.conexion = DBManager.getInstance().getConnection();
//            this.conexion.setAutoCommit(false);
//            String sql = this.generarSQLParaInsercion();
//            this.statement = this.conexion.prepareCall(sql);
//            this.statement.setString(1, almacen.getNombre());
//            this.statement.setInt(2, almacen.getAlmacen_central() ? 1 : 0);
//            //resultado = this.statement.executeUpdate();
//            this.statement.executeUpdate();
//            resultado = this.retornarUltimoAutoGenerado();
//            this.conexion.commit();
//        } catch (SQLException ex) {
//            System.err.println("Error al intentar insertar - " + ex);
//            try {
//              if (this.conexion != null) {
//                  this.conexion.rollback();
//              }
//            } catch (SQLException ex1) {
//              System.err.println("Error al hacer rollback - " + ex1);
//            }
//        } finally {
//            try {
//                if (this.conexion != null) {
//                    this.conexion.close();
//                }
//            } catch (SQLException ex) {
//                System.err.println("Error al cerrar la conexión - " + ex);
//            }
//        }
//        return resultado;
    }

    @Override
    public AlmacenesDTO obtenerPorId(Integer almacenId) {
        this.almacen = new AlmacenesDTO();
        this.almacen.setAlmacenId(almacenId);
        super.obtenerPorId();
        return this.almacen;
//        AlmacenesDTO almacen = null;
//        try {
//            this.conexion = DBManager.getInstance().getConnection();
//            String sql = this.generarSQLParaObtenerPorId();
//            this.statement = this.conexion.prepareCall(sql);
//            this.statement.setInt(1, almacenId);
//            this.resultSet = this.statement.executeQuery();
//            if (this.resultSet.next()) {
//                almacen = new AlmacenesDTO();
//                almacen.setAlmacenId(this.resultSet.getInt("ALMACEN_ID"));
//                almacen.setNombre(this.resultSet.getString("NOMBRE"));
//                almacen.setAlmacen_central(this.resultSet.getInt("ALMACEN_CENTRAL") == 1);
//            }
//        } catch (SQLException ex) {
//            System.err.println("Error al intentar obtenerPorId - " + ex);
//        } finally {
//            try {
//                if (this.conexion != null) {
//                    this.conexion.close();
//                }
//            } catch (SQLException ex) {
//                System.err.println("Error al cerrar la conexión - " + ex);
//            }
//        }
//        return almacen;
    }

    @Override
    public ArrayList<AlmacenesDTO> listarTodos() {
        return (ArrayList<AlmacenesDTO>) super.listarTodos();
//        ArrayList<AlmacenesDTO> listaAlmacenes = new ArrayList<>();
//        try {
//            this.conexion = DBManager.getInstance().getConnection();
//            String sql = this.generarSQLParaListarTodos();
//            this.statement = this.conexion.prepareCall(sql);
//            this.resultSet = this.statement.executeQuery();
//            while (this.resultSet.next()) {
//                AlmacenesDTO almacen = new AlmacenesDTO();
//                almacen.setAlmacenId(this.resultSet.getInt("ALMACEN_ID"));
//                almacen.setNombre(this.resultSet.getString("NOMBRE"));
//                almacen.setAlmacen_central(this.resultSet.getInt("ALMACEN_CENTRAL") == 1);
//                listaAlmacenes.add(almacen);
//            }
//        } catch (SQLException ex) {
//            System.err.println("Error al intentar listarTodos - " + ex);
//        } finally {
//            try {
//                if (this.conexion != null) {
//                    this.conexion.close();
//                }
//            } catch (SQLException ex) {
//                System.err.println("Error al cerrar la conexión - " + ex);
//            }
//        }
//        return listaAlmacenes;
    }

    @Override
    public Integer modificar(AlmacenesDTO almacen) {
        this.almacen = almacen;
        return super.modificar();
//        int resultado = 0;
//        try {
//            this.conexion = DBManager.getInstance().getConnection();
//            this.conexion.setAutoCommit(false);
//            String sql = this.generarSQLParaModificacion();
//            this.statement = this.conexion.prepareCall(sql);
//            this.statement.setString(1, almacen.getNombre());
//            this.statement.setInt(2, almacen.getAlmacen_central() ? 1 : 0);
//            this.statement.setInt(3, almacen.getAlmacenId());
//            resultado = this.statement.executeUpdate();
//            this.conexion.commit();
//        } catch (SQLException ex) {
//            System.err.println("Error al intentar modificar - " + ex);
//            try {
//              if (this.conexion != null) {
//                  this.conexion.rollback();
//              }
//            } catch (SQLException ex1) {
//              System.err.println("Error al hacer rollback - " + ex1);
//            }
//        } finally {
//            try {
//                if (this.conexion != null) {
//                    this.conexion.close();
//                }
//            } catch (SQLException ex) {
//                System.err.println("Error al cerrar la conexión - " + ex);
//            }
//        }
//        return resultado;
    }

    @Override
    public Integer eliminar(AlmacenesDTO almacen) {
        this.almacen = almacen;
        return super.eliminar();
//        int resultado = 0;
//        try {
//            this.conexion = DBManager.getInstance().getConnection();
//            this.conexion.setAutoCommit(false);
//            String sql = this.generarSQLParaEliminacion();
//            this.statement = this.conexion.prepareCall(sql);
//            this.statement.setInt(1, almacen.getAlmacenId());
//            resultado = this.statement.executeUpdate();
//            this.conexion.commit();
//        } catch (SQLException ex) {
//            System.err.println("Error al intentar eliminar - " + ex);
//            try {
//              if (this.conexion != null) {
//                  this.conexion.rollback();
//              }
//            } catch (SQLException ex1) {
//              System.err.println("Error al hacer rollback - " + ex1);
//            }
//        } finally {
//            try {
//                if (this.conexion != null) {
//                    this.conexion.close();
//                }
//            } catch (SQLException ex) {
//                System.err.println("Error al cerrar la conexión - " + ex);
//            }
//        }
//        return resultado;
    }

}

File: 0682/sesión 15  - Master Detail/sin Master Detail/SoftInv/SoftInvPersistance/DAOImpl/AlmacenDAOImpl.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Data.Common;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Xml.Linq;
using SoftInv.DAO;
using SoftInv.DAOImpl.Util;
using SoftInv.Model;
using static System.Collections.Specialized.BitVector32;

namespace SoftInv.DAOImpl
{
    public class AlmacenDAOImpl : DAOImplBase, AlmacenDAO
    {
        private AlmacenesDTO almacen;

        public AlmacenDAOImpl() : base("INV_ALMACENES")
        {
            this.retornarLlavePrimaria = true;
            this.almacen = null;
        }

        protected override void ConfigurarListaDeColumnas()
        {
            this.listaColumnas.Add(new Columna("ALMACEN_ID", true, true));
            this.listaColumnas.Add(new Columna("NOMBRE", false, false));
            this.listaColumnas.Add(new Columna("ALMACEN_CENTRAL", false, false));
        }

        protected override void IncluirValorDeParametrosParaInsercion()
        {
            AgregarParametro("@NOMBRE", this.almacen.Nombre);
            AgregarParametro("@ALMACEN_CENTRAL", (bool)this.almacen.AlmacenCentral ? 1 : 0);
        }
    
        protected override void IncluirValorDeParametrosParaModificacion()
        {            
            AgregarParametro("@ALMACEN_ID", this.almacen.AlmacenId);
            AgregarParametro("@NOMBRE", this.almacen.Nombre);
            AgregarParametro("@ALMACEN_CENTRAL", (bool)this.almacen.AlmacenCentral ? 1 : 0);
        }
    
        protected override void IncluirValorDeParametrosParaEliminacion()
        {
            AgregarParametro("@ALMACEN_ID", this.almacen.AlmacenId);
        }

        protected override void IncluirValorDeParametrosParaObtenerPorId()
        {
            AgregarParametro("@ALMACEN_ID", this.almacen.AlmacenId);
        }

        protected override void InstanciarObjetoDelResultSet(DbDataReader lector) 
        {
            this.almacen = new AlmacenesDTO();
            this.almacen.AlmacenId = this.lector.GetInt32(0);
            this.almacen.Nombre = this.lector.GetString(1);
            this.almacen.AlmacenCentral = this.lector.GetInt32(2) == 1;
        }

        protected override void LimpiarObjetoDelResultSet()
        {
            this.almacen = null;
        }

        protected override void AgregarObjetoALaLista(IList<Object> lista, DbDataReader lector)
        {
            this.InstanciarObjetoDelResultSet(lector);
            lista.Add(this.almacen);
        }

        public int Insertar(AlmacenesDTO almacen)
        {
            this.almacen = almacen;
            return base.Insertar();
        }

        public int Modificar(AlmacenesDTO almacen)
        {
            this.almacen = almacen;
            return base.Modificar();
        }

        public int Eliminar(AlmacenesDTO almacen)
        {
            this.almacen = almacen;
            return base.Eliminar();
        }

        public AlmacenesDTO ObtenerPorId(int almacenId)
        {
            this.almacen = new AlmacenesDTO();
            this.almacen.AlmacenId = almacenId;
            base.ObtenerPorId();
            return this.almacen;
        }

        public new IList<AlmacenesDTO> ListarTodos()
        {
            return base.ListarTodos()
               .Cast<AlmacenesDTO>()  
               .ToList();
        }
    }                    
}
